File: examples/double_pendulum/double_pendulum.cc
// A simple test case: simulation of a double inverted pendulum.
// There are no contact forces.
// This simulation checks the overall simulator sanity (i.e. conservation of energy) and genericity (working
// with something that is not an exoskeleton).

#include <iostream>

#include "jiminy/core/engine/Engine.h"
#include "jiminy/core/robot/BasicMotors.h"
#include "jiminy/core/control/ControllerFunctor.h"
#include "jiminy/core/io/FileDevice.h"
#include "jiminy/core/Utilities.h"
#include "jiminy/core/Types.h"

#include <boost/filesystem.hpp>


using namespace jiminy;

void computeCommand(float64_t                   const & t,
                    Eigen::Ref<vectorN_t const> const & q,
                    Eigen::Ref<vectorN_t const> const & v,
                    sensorsDataMap_t            const & sensorsData,
                    vectorN_t                         & u)
{
    // No controller: energy should be preserved.
    u.setZero();
}

void internalDynamics(float64_t                   const & t,
                      Eigen::Ref<vectorN_t const> const & q,
                      Eigen::Ref<vectorN_t const> const & v,
                      sensorsDataMap_t            const & sensorsData,
                      vectorN_t                         & u)
{
    u.setZero();
}

bool_t callback(float64_t const & t,
                vectorN_t const & q,
                vectorN_t const & v)
{
    return true;
}

int main(int argc, char_t * argv[])
{
    // =====================================================================
    // ==================== Extract the user paramaters ====================
    // =====================================================================

    // Set URDF and log output.
    boost::filesystem::path const filePath(__FILE__);
    auto const jiminySrcPath = filePath.parent_path().parent_path().parent_path();
    auto const urdfPath = jiminySrcPath / "data/double_pendulum/double_pendulum.urdf";
    auto const outputDirPath = boost::filesystem::temp_directory_path();

    // =====================================================================
    // ============ Instantiate and configure the simulation ===============
    // =====================================================================

    // Instantiate timer
    Timer timer;

    timer.tic();

    // Instantiate and configuration the robot
    std::vector<std::string> motorJointNames{"SecondPendulumJoint"};

    auto robot = std::make_shared<Robot>();
    configHolder_t modelOptions = robot->getModelOptions();
    boost::get<bool_t>(boost::get<configHolder_t>(modelOptions.at("joints")).at("positionLimitFromUrdf")) = true;
    boost::get<bool_t>(boost::get<configHolder_t>(modelOptions.at("joints")).at("velocityLimitFromUrdf")) = true;
    robot->setModelOptions(modelOptions);
    robot->initialize(urdfPath.string(), false);
    for (std::string const & jointName : motorJointNames)
    {
        auto motor = std::make_shared<SimpleMotor>(jointName);
        robot->attachMotor(motor);
        motor->initialize(jointName);
    }

    // Instantiate and configuration the controller
    auto controller = std::make_shared<ControllerFunctor<decltype(computeCommand),
                                                         decltype(internalDynamics)> >(computeCommand, internalDynamics);
    controller->initialize(robot.get());

    // Instantiate and configuration the engine
    auto engine = std::make_shared<Engine>();
    configHolder_t simuOptions = engine->getOptions();
    boost::get<bool_t>(boost::get<configHolder_t>(simuOptions.at("telemetry")).at("enableConfiguration")) = true;
    boost::get<bool_t>(boost::get<configHolder_t>(simuOptions.at("telemetry")).at("enableVelocity")) = true;
    boost::get<bool_t>(boost::get<configHolder_t>(simuOptions.at("telemetry")).at("enableAcceleration")) = true;
    boost::get<bool_t>(boost::get<configHolder_t>(simuOptions.at("telemetry")).at("enableEffort")) = true;
    boost::get<bool_t>(boost::get<configHolder_t>(simuOptions.at("telemetry")).at("enableEnergy")) = true;
    boost::get<vectorN_t>(boost::get<configHolder_t>(simuOptions.at("world")).at("gravity"))(2) = -9.81;
    boost::get<std::string>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("odeSolver")) = std::string("runge_kutta_dopri5");
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("tolRel")) = 1.0e-5;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("tolAbs")) = 1.0e-4;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("dtMax")) = 3.0e-3;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("dtRestoreThresholdRel")) = 0.2;
    boost::get<int32_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("iterMax")) = 100000U; // -1 to disable
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("timeout")) = -1; // -1 to disable
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("sensorsUpdatePeriod")) = 1.0e-3;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("controllerUpdatePeriod")) = 1.0e-3;
    boost::get<bool_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("logInternalStepperSteps")) = false;
    boost::get<uint32_t>(boost::get<configHolder_t>(simuOptions.at("stepper")).at("randomSeed")) = 0U; // Use time(nullptr) for random seed.
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("contacts")).at("stiffness")) = 1e6;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("contacts")).at("damping")) = 2000.0;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("contacts")).at("frictionDry")) = 5.0;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("contacts")).at("frictionViscous")) = 5.0;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("contacts")).at("frictionStictionVel")) = 0.01;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("contacts")).at("frictionStictionRatio")) = 0.5;
    boost::get<float64_t>(boost::get<configHolder_t>(simuOptions.at("contacts")).at("transitionEps")) = 0.001;
    engine->setOptions(simuOptions);
    engine->initialize(robot, controller, callback);

    timer.toc();

    // =====================================================================
    // ======================= Run the simulation ==========================
    // =====================================================================

    // Prepare options
    Eigen::VectorXd x0 = Eigen::VectorXd::Zero(4);
    x0(1) = 0.1;
    float64_t const tf = 3.0;

    // Run simulation
    timer.tic();
    engine->simulate(tf, x0);
    timer.toc();
    std::cout << "Simulation time: " << (timer.dt * 1.0e3) << "ms" << std::endl;

    // Write the log file
    std::vector<std::string> header;
    matrixN_t log;
    engine->getLogData(header, log);
    std::cout << log.rows() << " log points" << std::endl;
    std::cout << engine->getStepperState().iter << " internal integration steps" << std::endl;
    engine->writeLogTxt((outputDirPath / "log.txt").string());
    engine->writeLogBinary((outputDirPath / "log.data").string());

    return 0;
}

File: python/jiminy_py/examples/double_pendulum.py
import os
import tempfile
import time
import numpy as np
import matplotlib.pyplot as plt

import jiminy_py.core as jiminy
from jiminy_py.viewer import play_logs_data


if __name__ == '__main__':
    # ######################## User parameters ################################

    script_dir = os.path.dirname(os.path.realpath(__file__))
    mesh_root_dir = os.path.join(script_dir, "../../../data/toys_models")
    urdf_path = os.path.join(
        mesh_root_dir, "double_pendulum/double_pendulum.urdf")

    # ################### Initialize the simulation ###########################

    # Instantiate the robot
    motor_joint_names = ("SecondPendulumJoint",)
    robot = jiminy.Robot()
    robot.initialize(
        urdf_path, has_freeflyer=False, mesh_package_dirs=[mesh_root_dir])
    for joint_name in motor_joint_names:
        motor = jiminy.SimpleMotor(joint_name)
        robot.attach_motor(motor)
        motor.initialize(joint_name)

    # Define a controller doing nothing and instantiate it
    class Controller(jiminy.BaseController):
        def compute_command(self, t, q, v, command):
            pass

        def internal_dynamics(self, t, q, v, u_custom):
            pass

    robot.controller = Controller()

    # Instantiate the engine
    engine = jiminy.Engine()
    engine.add_robot(robot)

    # ################## Configuration the simulation #########################

    robot_options = robot.get_options()
    engine_options = engine.get_options()

    robot_options["telemetry"]["enableImuSensors"] = True
    engine_options["telemetry"]["isPersistent"] = True
    engine_options["telemetry"]["logInternalStepperSteps"] = False
    engine_options["telemetry"]["enableConfiguration"] = True
    engine_options["telemetry"]["enableVelocity"] = True
    engine_options["telemetry"]["enableAcceleration"] = True
    engine_options["telemetry"]["enableEffort"] = True
    engine_options["telemetry"]["enableForceExternal"] = False
    engine_options["telemetry"]["enableCommand"] = True
    engine_options["telemetry"]["enableEnergy"] = True
    engine_options["world"]["gravity"][2] = -9.81
    engine_options["stepper"]["odeSolver"] = "runge_kutta_dopri"  # ["runge_kutta_dopri", "euler_explicit"]
    engine_options["stepper"]["tolRel"] = 1.0e-5
    engine_options["stepper"]["tolAbs"] = 1.0e-4
    engine_options["stepper"]["dtMax"] = 2.0e-3  # 2.0e-4 for "euler_explicit", 3.0e-3 for "runge_kutta_dopri"
    engine_options["stepper"]["sensorsUpdatePeriod"] = 1.0e-3
    engine_options["stepper"]["controllerUpdatePeriod"] = 1.0e-3
    engine_options["stepper"]["randomSeedSeq"] = np.array([0,], dtype=np.uint32)
    engine_options['contacts']['model'] = "spring_damper"
    engine_options['contacts']['stiffness'] = 1.0e6
    engine_options['contacts']['damping'] = 2000.0
    engine_options['contacts']['friction'] = 5.0
    engine_options['contacts']['transitionEps'] = 0.001
    engine_options['contacts']['transitionVelocity'] = 0.01

    robot.set_options(robot_options)
    engine.set_options(engine_options)

    # ####################### Run the simulation ##############################

    q0 = np.zeros((2,))
    q0[1] = 0.1
    v0 = np.zeros((2,))
    tf = 3.0

    start = time.time()
    engine.simulate(tf, q0, v0)
    end = time.time()
    print(f"Simulation time: {(end - start) * 1.0e3:03.0f}ms")

    # ####################### Extract the results #############################

    log_data = engine.log_data
    log_vars, log_constants = log_data["variables"], log_data["constants"]
    print(f"{log_vars['Global.Time'].size} log points")
    print(log_constants)

    # Save the log in HDF5
    engine.write_log(
        os.path.join(tempfile.gettempdir(), "log.hdf5"), format="hdf5")

    # ###################### Display the results ##############################

    # Plot some data using standard tools only
    plt.figure()
    plt.plot(log_vars['Global.Time'], log_vars['energy'])
    plt.title('Double pendulum energy')
    plt.grid()
    plt.show()

    # Replay the simulation
    play_logs_data(log_data,
                   robot,
                   speed_ratio=0.5,
                   camera_pose=((0.0, 7.0, 0.0), (np.pi/2, 0.0, np.pi), None),
                   delete_robot_on_close=False)
